File: examples/loadtest/latency/c/latency.c
/*
 * Per-op latency: timed push + pop_into pair.
 * Env: EMQ_LOAD_N, EMQ_LOAD_PAYLOAD, EMQ_LOAD_WARMUP
 * Prints: LATENCY lang=c payload=N n=N p50_ns=... p99_ns=... p999_ns=... p9999_ns=...
 */
#include "emq/emq.h"

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>

static int env_int(const char *key, int def) {
  const char *v = getenv(key);
  if (!v || !*v) return def;
  return atoi(v);
}

static uint64_t now_ns(void) {
  struct timespec ts;
  clock_gettime(CLOCK_MONOTONIC, &ts);
  return (uint64_t)ts.tv_sec * 1000000000ULL + (uint64_t)ts.tv_nsec;
}

static int cmp_u64(const void *a, const void *b) {
  uint64_t da = *(const uint64_t *)a;
  uint64_t db = *(const uint64_t *)b;
  return (da > db) - (da < db);
}

static uint64_t pct(uint64_t *sorted, size_t n, double p) {
  if (n == 0) return 0;
  size_t idx = (size_t)(p * (double)(n - 1) + 0.5);
  if (idx >= n) idx = n - 1;
  return sorted[idx];
}

int main(void) {
  int n = env_int("EMQ_LOAD_N", 1000000);
  int payload_len = env_int("EMQ_LOAD_PAYLOAD", 64);
  int warmup = env_int("EMQ_LOAD_WARMUP", 50000);
  uint32_t capacity = 4096;
  uint8_t *payload, *dst;
  uint64_t *lat;
  emq_runtime *rt = NULL;
  emq_queue *q = NULL;
  emq_queue_opts opts;
  size_t got;
  int i;

  if (n < 1) n = 1;
  payload = (uint8_t *)malloc((size_t)payload_len);
  dst = (uint8_t *)malloc((size_t)payload_len);
  lat = (uint64_t *)malloc(sizeof(uint64_t) * (size_t)n);
  if (!payload || !dst || !lat) {
    fprintf(stderr, "oom\n");
    return 1;
  }
  memset(payload, 0xAB, (size_t)payload_len);

  if (emq_runtime_create(&rt) != EMQ_OK) return 1;
  emq_queue_opts_default(&opts);
  opts.storage = EMQ_STORAGE_FAST;
  opts.policy = EMQ_POLICY_FIFO;
  opts.capacity = capacity;
  opts.producers = 1;
  opts.consumers = 1;
  if (emq_queue_create(rt, "lat-c", &opts, &q) != EMQ_OK) return 1;

  for (i = 0; i < warmup; ++i) {
    if (emq_push(q, payload, (size_t)payload_len, NULL) != EMQ_OK) return 1;
    if (emq_pop_into(q, dst, (size_t)payload_len, &got, NULL, 0) != EMQ_OK)
      return 1;
  }

  for (i = 0; i < n; ++i) {
    uint64_t t0 = now_ns();
    if (emq_push(q, payload, (size_t)payload_len, NULL) != EMQ_OK) return 1;
    if (emq_pop_into(q, dst, (size_t)payload_len, &got, NULL, 0) != EMQ_OK)
      return 1;
    lat[i] = now_ns() - t0;
  }

  qsort(lat, (size_t)n, sizeof(uint64_t), cmp_u64);
  printf("LATENCY lang=c payload=%d n=%d p50_ns=%llu p99_ns=%llu p999_ns=%llu "
         "p9999_ns=%llu\n",
         payload_len, n, (unsigned long long)pct(lat, (size_t)n, 0.50),
         (unsigned long long)pct(lat, (size_t)n, 0.99),
         (unsigned long long)pct(lat, (size_t)n, 0.999),
         (unsigned long long)pct(lat, (size_t)n, 0.9999));

  emq_queue_close(q);
  emq_runtime_destroy(rt);
  free(payload);
  free(dst);
  free(lat);
  return 0;
}

File: examples/loadtest/latency/latency.rs
use emq::{QueueOpts, Runtime};
use std::env;
use std::time::Instant;

fn env_usize(key: &str, def: usize) -> usize {
    env::var(key)
        .ok()
        .and_then(|s| s.parse().ok())
        .unwrap_or(def)
}

fn pct(sorted: &[u64], p: f64) -> u64 {
    if sorted.is_empty() {
        return 0;
    }
    let idx = (p * (sorted.len() - 1) as f64 + 0.5) as usize;
    sorted[idx.min(sorted.len() - 1)]
}

fn main() -> emq::Result<()> {
    let n = env_usize("EMQ_LOAD_N", 1_000_000);
    let payload_len = env_usize("EMQ_LOAD_PAYLOAD", 64);
    let warmup = env_usize("EMQ_LOAD_WARMUP", 50_000);
    let payload = vec![0xABu8; payload_len];
    let mut dst = vec![0u8; payload_len];
    let mut lat = vec![0u64; n];

    let rt = Runtime::new()?;
    let q = rt.create_queue(
        "lat-rs",
        Some(QueueOpts {
            capacity: 4096,
            producers: 1,
            consumers: 1,
        }),
    )?;

    for _ in 0..warmup {
        q.push(&payload)?;
        let _ = q.pop_into(&mut dst, None)?;
    }

    for i in 0..n {
        let t0 = Instant::now();
        q.push(&payload)?;
        let _ = q.pop_into(&mut dst, None)?;
        lat[i] = t0.elapsed().as_nanos() as u64;
    }

    lat.sort_unstable();
    println!(
        "LATENCY lang=rust payload={payload_len} n={n} p50_ns={} p99_ns={} p999_ns={} p9999_ns={}",
        pct(&lat, 0.50),
        pct(&lat, 0.99),
        pct(&lat, 0.999),
        pct(&lat, 0.9999)
    );
    Ok(())
}
